src/analytics/analytics.js

```diff
@@ -41,25 +45,69 @@ export class Analytics {
         requestAnimationFrame(this.draw);
     }
 
-    add() {
-        if (this.added) {
-            return;
+    /**
+     * On envelope open
+     * add, load pinned region cards, load spaghetti, set timeline
+     */
+    open(frameId) {
+        if (this.opened) {
+            if (this.activeEnvelope === frameId) {
+                setActiveFrame(frameId);
+                return;
+            }
+            if (this.activeEnvelope) {
+                this.blur(this.activeEnvelope);
+            }
         }
-        this.added = true;
+        this.activeEnvelope = frameId;
+        this.opened = true;
 
         this.createNewPinnedRegionCardsContainer();
         document.body.appendChild(this.container);
+        this.container.style.display = 'block';
+        setActiveFrame(frameId);
         showAnalyzerSettingsUI();
     }
 
-    remove() {
-        if (!this.added) {
+    /**
+     * On envelope close
+     * remove pinned region cards, remove timeline, remove spaghetti
+     */
+    close(frameId) {
+        if (frameId !== this.activeEnvelope) {
+            return;
+        }
+
+        if (!this.opened) {
+            return;
+        }
+
+        this.opened = false;
+
+        this.blur(frameId);
+        clearHistoricalData(frameId);
+    }
+
+    /**
+     * On envelope focus (unblur)
+     */
+    focus(frameId) {
+        this.open(frameId);
+    }
+
+    /**
+     * On envelope blur
+     * Remove all 2d ui
+     */
+    blur(frameId) {
```

Comment:
I understand  focus/blur naming but at least blur is confusing when encountered in the code somewhere. Maybe foreground/background is clearer pairs but they are not verbs.

Reply:
I have not realised that it is actually used more broadly. It is probably not worth refactoring before Liveworx.  However, it may be an improvement task in backlog to see if better naming can be found.